// @ytdl/ytdl/dist/src/utils/logger.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
exports.__esModule = true;
var winston_1 = __importDefault(require("winston"));
/**
 * Returns a new logger with transports to console.
 * @param level Stores the logger level
 */
function createLogger(level) {
    var levels = {
        error: 0,
        warn: 1,
        info: 2,
        http: 3,
        verbose: 4,
        debug: 5,
        silly: 6
    };
    if (levels[level] === undefined) {
        throw new Error('Invalid logging level.');
    }
    var logger = winston_1["default"].createLogger({
        level: level,
        transports: [
            new winston_1["default"].transports.Console({
                level: level
            }),
        ],
        format: winston_1["default"].format.combine(winston_1["default"].format.colorize(), winston_1["default"].format.simple())
    });
    return logger;
}
exports.createLogger = createLogger;
/**
 * Stores the default logger.
 */
var logger = winston_1["default"].createLogger({
    transports: [
        new winston_1["default"].transports.Console({
            level: 'info'
        }),
    ],
    format: winston_1["default"].format.combine(winston_1["default"].format.colorize(), winston_1["default"].format.simple())
});
exports["default"] = logger;
